ft232/mpsse.py

```diff
@@ -0,0 +1,555 @@
+        return 0x80 | (ConfigHighLowPins << 1) | SetGetPinConfig
+
+    def enableLoopback(self):
+        """ When Loopback is enabled TDI/DO and TDO/DI pins are connected with eachother internally to transfer data without an external device. This is for testing.
+        This function implements chapter 3.7 - Loopback Commands.
+        It does not return anything but directly writes the Commandbyte to the device.
+        """
+        self.__Ft232.write(bytes([0x84]))
+
+    def disableLoopback(self):
+        """ This disables Loopback. See enableLoopback() for mor details
+        This function implements chapter 3.7 - Loopback Commands.
+        It does not return anything but directly writes the Commandbyte to the device.
+        """
+        self.__Ft232.write(bytes([0x85]))
+
+    def __setClockDivisor(self, Divisor):
+        """ The Clock of the MPSSE device can be changed by setting a divisor, which lowers the Baseclock of the device by following equation:
+        Clock = Baseclock / (( 1 + Divisor ) * 2)
+        This should not be called directly. Instead use _setClock().
+        This function implements chapter 3.8 - Clock Divisor.
+        It does not return anything but directly writes the Commandbyte and new Divisor to the device.
+        """
+        DivisorLowByte = Divisor & 0xff
+        DivisorHighByte = (Divisor >> 8) & 0xff
+        print("clock divisor set to " + str(Divisor))
+        self.__Ft232.write(bytes([0x86, DivisorLowByte, DivisorHighByte]))
+
+    def _cpuMode(self): #feel free to change the name
+        """ To be implemented.
+        See chapter 4 - Instructions for CPU mode for more details
+        """
+        pass
+
+    def _mcuHost(self): #feel free to change the name
+        """ To be implemented.
+        See chapter 5 - Instructions for use in both MPSSE and MCU Host Emulation Modes for more details
+        """
+        pass
+
+    def __enableClockDivide(self):
+        """ This will turn on the Baseclock devide of the MPSSE unit by 5.
+        This function implements chapter 6.2. Enable Clk Divide by 5.
+        It does not return anything but directly writes the Commandbyte and new Divisor to the device.
+        """
+        if self.__isModernDevice():
+            self.__Ft232.write(bytes([0x8b]))
+
+    def __disableClockDivide(self):
+        """ This turn off the Baseclock devide of the MPSSE unit by 5.
+        This function implements chapter 6.1. Disable Clk Divide by 5.
+        It does not return anything but directly writes the Commandbyte and new Divisor to the device.
+        """
+        if self.__isModernDevice():
+            self.__Ft232.write(bytes([0x8a]))
+
+    def __enableThreePhaseClocking(self):
+        """ Missing description.
+        This function implements chapter 6.3 Enable 3 Phase Data Clocking.
+        It does not return anything but directly writes the Commandbyte and new Divisor to the device.
+        """
+        if self.__isModernDevice():
```

Comment:
This pattern makes me nervous. So if it's not a modern device then nothing happens? I think setClock() should be a bit more careful here, if the user requests Three Phase Clocking on a device that doesn't support it, there should be an error; not silent acceptance and perhaps breaking things further (by multiplying the frequency by 1.5 when it probably doesn't apply).

Reply:
So i should raise exceptions here right?

Reply:
Yes, but given the way you have the code structured now that's probably not going to work because these functions get called no matter what and therefore will always raise an exception if the hardware doesn't support it.